src/System.Net.Http/src/uap/System/Net/HttpHandlerToFilter.cs

```diff
@@ -26,65 +26,207 @@ namespace System.Net.Http
+        private readonly RTHttpBaseProtocolFilter _filter;
+        private readonly RTHttpBaseProtocolFilter _filterWithNoCredentials;
         private int _filterMaxVersionSet;
+        private HttpClientHandler _handler;
 
-        internal HttpHandlerToFilter(RTHttpBaseProtocolFilter filter)
+        internal HttpHandlerToFilter(
+            RTHttpBaseProtocolFilter filter,
+            RTHttpBaseProtocolFilter filterWithNoCredentials,
+            HttpClientHandler handler)
         {
             if (filter == null)
             {
                 throw new ArgumentNullException(nameof(filter));
             }
 
-            _next = filter;
+            _filter = filter;
+            _filterWithNoCredentials = filterWithNoCredentials;
             _filterMaxVersionSet = 0;
+            _handler = handler;
         }
 
         internal string RequestMessageLookupKey { get; set; }
         internal string SavedExceptionDispatchInfoLookupKey { get; set; }
 
         protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancel)
         {
+            int redirects = 0;
+            HttpMethod requestHttpMethod;
+            bool skipRequestContentIfPresent = false;
+            HttpResponseMessage response = null;
+
             if (request == null)
             {
                 throw new ArgumentNullException(nameof(request));
             }
-            cancel.ThrowIfCancellationRequested();
 
-            RTHttpRequestMessage rtRequest = await ConvertRequestAsync(request).ConfigureAwait(false);
+            requestHttpMethod = request.Method;
 
-            RTHttpResponseMessage rtResponse;
-            try
-            {
-                rtResponse = await _next.SendRequestAsync(rtRequest).AsTask(cancel).ConfigureAwait(false);
-            }
-            catch (TaskCanceledException)
-            {
-                throw;
-            }
-            catch (Exception)
+            while (true)
             {
-                object info;
-                if (rtRequest.Properties.TryGetValue(SavedExceptionDispatchInfoLookupKey, out info))
+                cancel.ThrowIfCancellationRequested();
+
+                if (response != null)
                 {
-                    ((ExceptionDispatchInfo)info).Throw();
+                    response.Dispose();
+                    response = null;
                 }
 
-                throw;
-            }
+                RTHttpRequestMessage rtRequest = await ConvertRequestAsync(
+                    request,
+                    requestHttpMethod,
+                    skipRequestContentIfPresent).ConfigureAwait(false);
+
+                RTHttpResponseMessage rtResponse;
+                try
+                {
+                    if (redirects > 0)
+                    {
+                        rtResponse = await _filterWithNoCredentials.SendRequestAsync(rtRequest).AsTask(cancel).ConfigureAwait(false);
```

Comment:
I suppose I could use a Lazy<T> pattern here and only create the rtFilterWithNoCredentials as a "clone, minus credentials" when we first need it.  At that point, since the first operation on the rtFilter has happened, the original rtFilter properties are considered locked. Therefore, we don't have to constantly re-create the rtFilterWithNoCredentials.

I thought it was simpler to just create it and update properties but perhaps I can revise that.